rustls/src/crypto/hpke.rs

```diff
@@ -0,0 +1,131 @@
+/// [HPKE ciphersuite profile]: <https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-17#section-9>
+pub trait HpkeProvider: Debug + Send + Sync + 'static {
+    /// Create an HPKE instance for the given mode, KEM, KDF and AEAD.
+    ///
+    /// May return an error if the mode, KEM, KDF or AEAD are unsupported by the HPKE provider.
+    fn setup(
+        &self,
+        mode: Mode,
+        kem: HpkeKem,
+        kdf: HpkeKdf,
+        aead: HpkeAead,
+    ) -> Result<Box<dyn Hpke>, Error>;
+
+    /// Does the provider support the given mode?
+    fn supports_mode(&self, mode: Mode) -> bool;
+
+    /// Does the provider support the given [HpkeKem]?
+    fn supports_kem(&self, kem: HpkeKem) -> bool;
+
+    /// Does the provider support the given [HpkeKdf]?
+    fn supports_kdf(&self, kdf: HpkeKdf) -> bool;
+
+    /// Does the provider support the given [HpkeAead]?
+    fn supports_aead(&self, aead: HpkeAead) -> bool;
+}
+
+/// Available HPKE authentication modes.
+///
+/// The base case for HPKE is unauthenticated encryption to a public key. In addition to this
+/// mode there are three modes to provide a form of authentication.
+///
+/// See [RFC 9180 Section 5](https://www.rfc-editor.org/rfc/rfc9180.html#name-hybrid-public-key-encryptio)
+/// for more details.
+#[derive(PartialEq, Copy, Clone, Debug)]
+#[repr(u8)]
+pub enum Mode {
+    /// Encryption to a public key, no additional authentication.
+    Base = 0x00,
+
+    /// Encryption to a public key, authenticated with a pre-shared secret key.
+    Psk = 0x01,
+
+    /// Encryption to a public key, authenticated with the sender's private key.
+    Auth = 0x02,
+
+    /// Encryption to a public key, authenticated with the sender's private key and a
+    /// pre-shared secret key.
+    AuthPsk = 0x03,
+}
+
+/// An HPKE instance that can produce sender and receiver contexts.
+pub trait Hpke: Debug + Send + Sync {
+    /// Create a sender context to encrypt messages to a receiver.
+    ///
+    /// This will return an encapsulated secret and a sender context. The sender context
+    /// can be used to seal messages to the receiver.
+    fn setup_sender(
+        &self,
+        pk_r: &HpkePublicKey,
+        info: &[u8],
+        psk: Option<&HpkePreSharedKey>,
+        psk_id: Option<&[u8]>,
+        sk_s: Option<&HpkePrivateKey>,
+    ) -> Result<(EncapsulatedSecret, Box<dyn HpkeSender + '_>), Error>;
+
+    /// Create a receiver context to decrypt messages from a sender.
+    fn setup_receiver(
+        &self,
+        enc: &EncapsulatedSecret,
+        sk_r: &HpkePrivateKey,
+        info: &[u8],
+        psk: Option<&HpkePreSharedKey>,
+        psk_id: Option<&[u8]>,
+        pk_s: Option<&HpkePublicKey>,
+    ) -> Result<Box<dyn HpkeReceiver + '_>, Error>;
+}
+
+/// An HPKE Sender context.
+///
+/// This is a stateful object that can be used to seal messages for receipt by
+/// a receiver.
+pub trait HpkeSender: Debug + Send + Sync {
```

Comment:
I would be minded to start with a one-shot interface for the sake of simplicity. Then move to a streaming interface once everything is up and running?

That would obviate `HpkeSender` and `HpkeReceiver`, and make the methods in `trait Hpke` something like:

```rust
fn open(&self, 
        pk_r: &HpkePublicKey,
        info: &[u8],
        aad; &[u8],
        pt: &[u8],
) -> Result<(EncapsulatedSecret, Vec<u8>), Error>;
```

Reply:
Switched to one-shot.

Reply:
> I would be minded to start with a one-shot interface for the sake of simplicity.

Leaving a note here: we _do_ need the non-one-shot interface for ECH: specifically for handling HelloRetryRequests.

[Section 6.1.5 "Handshaking with ClientHelloInner"](https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-17#section-6.1.5) says:
> It encrypts EncodedClientHelloInner as described in [Section 6.1.1](https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-17#encrypting-clienthello), using the second partial ClientHelloOuterAAD, to obtain a second ClientHelloOuter. It reuses the original HPKE encryption context computed in [Section 6.1](https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-17#real-ech) and uses the empty string for enc.
> 
> The HPKE context maintains a sequence number, so this operation internally uses a fresh nonce for each AEAD operation. Reusing the HPKE context avoids an attack described in [Section 10.11.2](https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-17#flow-hrr-hijack).

So we specifically need to maintain the HPKE context used for the original client hello so that we can re-use it when creating the updated extension for a HRR response.

I'm trying to bundle up some required changes and will make sure this is one of them.